packages/common/src/message-input/message-input.tsx

```diff
@@ -0,0 +1,182 @@
+  placeholder?: string;
+  /** Option to set a draft message to display in the text window. */
+  draftMessage?: string;
+  /** Option to attach sender data directly to each message. Enable it for high-throughput environments.
+   * This is an alternative to providing a full list of users directly into the Chat provider. */
+  senderInfo?: boolean;
+  /** Option to enable/disable firing the typing events when a user is typing a message. */
+  typingIndicator?: boolean;
+  /** Option to enable/disable the internal file upload capability. */
+  fileUpload?: "image" | "all";
+  /** Option to disable the input from composing and sending messages. */
+  disabled?: boolean;
+  /** Option to hide the Send button. */
+  hideSendButton?: boolean;
+  /** Custom UI component to override default display for the Send button. */
+  sendButton?: JSX.Element | string;
+  /** Callback to handle an event when the text value changes. */
+  onChange?: (value: string) => void;
+  /** Callback to modify message content before sending it. This only works for text messages, not files. */
+  onBeforeSend?: (value: MessagePayload) => MessagePayload;
+  /** Callback for extra actions after sending a message. */
+  onSend?: (value: MessagePayload | File) => void;
+  /** Option to provide an extra actions renderer to add custom action buttons to the input. */
+  extraActionsRenderer?: () => JSX.Element;
+}
+
+/**
+ * Allows users to compose messages using text and emojis
+ * and automatically publish them on PubNub channels upon sending.
+ */
+export const useMessageInputCore = (props: CommonMessageInputProps) => {
+  const pubnub = usePubNub();
+
+  const [text, setText] = useState(props.draftMessage || "");
+  const [file, setFile] = useState<File>(null);
+  const [typingIndicatorSent, setTypingIndicatorSent] = useState(false);
+  const [loader, setLoader] = useState(false);
+  const [users] = useAtom(UsersMetaAtom);
+  const [theme] = useAtom(ThemeAtom);
+  const [channel] = useAtom(CurrentChannelAtom);
+  const [onErrorObj] = useAtom(ErrorFunctionAtom);
+  const onError = onErrorObj.function;
+  const [typingIndicatorTimeout] = useAtom(TypingIndicatorTimeoutAtom);
+
+  /*
+  /* Helper functions
+  */
+  const isValidInputText = () => {
+    return !!text.trim().length;
+  };
+
+  /*
+  /* Commands
+  */
+
+  const sendMessage = async () => {
+    try {
+      if (!file && !isValidInputText()) return;
+      let message = {
+        id: uuid.v4(),
+        text: file ? "" : text,
+        type: file ? "" : "default",
+        ...(props.senderInfo && { sender: users.find((u) => u.id === pubnub.getUUID()) }),
+        createdAt: new Date().toISOString(),
+      } as MessagePayload;
+      setLoader(true);
+
+      if (file) {
+        await pubnub.sendFile({ channel, file, message });
+        props.onSend && props.onSend(file);
+      } else if (text) {
+        if (props.onBeforeSend) message = props.onBeforeSend(message) || message;
```

Comment:
Just a question:
Maybe instead of constantly making sure that such and such an optional function was passed from the customer we could just define empty functions as defaultProps?

Reply:
We might try, not sure if I haven't tried this already and had some problems with this approach